src/core/machine-id-setup.c

```diff
@@ -107,26 +107,31 @@ int machine_id_setup(const char *root, sd_id128_t machine_id, sd_id128_t *ret) {
                         fd = open(etc_machine_id, O_RDONLY|O_CLOEXEC|O_NOCTTY);
                         if (fd < 0) {
                                 if (old_errno == EROFS && errno == ENOENT)
-                                        return log_error_errno(errno,
-                                                  "System cannot boot: Missing /etc/machine-id and /etc is mounted read-only.\n"
-                                                  "Booting up is supported only when:\n"
-                                                  "1) /etc/machine-id exists and is populated.\n"
-                                                  "2) /etc/machine-id exists and is empty.\n"
-                                                  "3) /etc/machine-id is missing and /etc is writable.\n");
+                                        log_error_errno(errno,
+                                                        "System cannot boot: Missing /etc/machine-id and /etc is mounted read-only.\n"
+                                                        "Booting up is supported only when:\n"
+                                                        "1) /etc/machine-id exists and is populated.\n"
+                                                        "2) /etc/machine-id exists and is empty.\n"
+                                                        "3) /etc/machine-id is missing and /etc is writable.\n");
```

Comment:
not sure i grok this, why make this non-fatal?

Do you intend here to allow the system to boot up with /etc/machine-id not set up? urks... I'd always say we should insist in /etc/machine-id being readable at any time any process that is not PID 1 runs...

i.e. that means insisting that /etc/machine-id exists or can be created by us. And if it exists as symlink but points elsewhere, then operate on that. i.e. issue chase_symlinks(CHASE_NONEXISTENT) to determine where the symlink (or chain of it) points, and then create/overmount things there.

Reply:
I'm not making it non-fatal, i'm just making sure that /run/machine-id is created whenever we can't read a correct value from /etc/machine-id.

Currently, the boot continues after this message, so I'm not changing anything really...

The problem with chasing symlink, is that it might points to a /var that is not mounted yet. In that case we can't create anything at the target location (since it's read-only) and we can't bind-mount on top of /etc/machine-id (since we can't mount on top of a dangling symlink) so we are screwed anyway. I'm opened to other ideas on how to deal with that case.. 

We could "completely break" things here and refuse to boot entirely, or (what I am trying to do) improve things and attempt to commit the machine-id later anyway. So it's setup for the next boot. 

Again, we currently boot after that scary warning, so the only change I do is making the commit of machine-id work correctly if it's possible when the commit service is invoked.

(arguably, i'm a bit in the dark here about what the "correct thing" to do is. I want to support symlink to /var, but having /var mounted be mandatory before systemd is started is good enough for my use-case. having it mounted later is just a bonus)

Reply:
Hmm, then change the message maybe? Saying "System cannot boot" and then booting is not very consitent.

Reply:
yes, i'll rework it even more. Adding a new correct case (symlink)